the long awaited kit commands

arnokeesman opened the pull request:
just adds a string list to the kits
`@p` is replaced with their username so it can work with LuckPerms and other stuff as well

arnokeesman committed:
fix running in dev env

arnokeesman committed:
feat: kit commands

John-Paul-R commented on src/main/java/dev/jpcode/kits/Kit.java:
```diff
@@ -16,16 +19,18 @@ public class Kit {
     private final KitInventory inventory;
     private final long cooldown;
     private @Nullable Item displayItem;
+    private @Nullable LinkedList<String> commands;
```
nit: recommend optimizing for read (rather than write) performance, as I generally assume that the number of times a kit is claimed will be >> the number of times the kit is initialized/configured.

```suggestion
    private @Nullable ArrayList<String> commands;
```

John-Paul-R replied:
Quick example of where this comes into play: on `KitCommandsManagerCommand.java:35`

```java
            for (int i = 1; i <= commands.size(); i++) {
                String command = commands.get(i - 1);
```

Since you're currently using a `LinkedList` this loop is actually O(n^2), because indexed lookups (`commands.get(idx)`) are `O(n)` for linked lists. ([the chart of gospel](https://www.bigocheatsheet.com/))

arnokeesman replied:
Thanks for the advice, was just using a List which doesn't have `.remove(o)` when you go from the Array/Stream to a list using `.asList`.
Replaced this with the first thing I encountered on the internet that did work

John-Paul-R commented on src/main/java/dev/jpcode/kits/Kit.java:
```diff
@@ -40,26 +45,50 @@ public Optional<Item> displayItem() {
     }
 
-    private static class StorageKey {
+    public Optional<LinkedList<String>> commands() {
+        return Optional.ofNullable(commands);
+    }
+
+    public boolean addCommand(String command) {
+        if (commands != null && commands.contains(command)) return false;
+        if (commands == null) this.commands = new LinkedList<>();
+        this.commands.add(command);
+        return true;
+    }
+
+    public boolean removeCommand(String command) {
+        if (commands().isEmpty() || !Objects.requireNonNull(commands).contains(command)) return false;
+        if (commands.size() == 1) commands = null;
+        else commands.remove(command);
+        return true;
+    }
```
While we do save a list allocation for the "no commands" case by making the list `@Nullable`, I think you'll find that a bunch of this code is simplified not-insignificantly by making it non-nullable (defaulted to empty list). (this simplicity is something I've come to value, less room to miss things, easier to read)

Its very appealing to me, for example, to not have to worry about whether the backing collection is null whenever we want to read from or write to it:

```suggestion
    public List<String> commands() {
        return commands;
    }

    public boolean addCommand(String command) {
        if (commands.contains(command)) return false;
        this.commands.add(command);
        return true;
    }

    public boolean removeCommand(String command) {
        if (commands().isEmpty() || !commands.contains(command)) return false;
        else commands.remove(command);
        return true;
    }
```

John-Paul-R replied:
I suppose this piece of guidance is "if `null` is supposed to behave identically to an empty list, use an empty list"

John-Paul-R commented on src/main/java/dev/jpcode/kits/command/KitCommandsManagerCommand.java:
```diff
@@ -0,0 +1,99 @@
+            for (int i = 1; i <= commands.size(); i++) {
+                String command = commands.get(i - 1);
+                message.append(Text.literal(String.format("\n#%d: %s", i, command))
+                    .setStyle(Style.EMPTY.withClickEvent(new ClickEvent(ClickEvent.Action.SUGGEST_COMMAND,
```
I didn't know about `SUGGEST_COMMAND`, somehow. Or I forgot. Very cool, I like it here.

John-Paul-R commented on src/main/java/dev/jpcode/kits/command/KitCommandsManagerCommand.java:
```diff
@@ -0,0 +1,99 @@
+        if (kit == null) return 0;
+
+        String command = StringArgumentType.getString(context, "command")
+            .replaceFirst("^/+", ""); // remove any slashes at the start
```
This will land you in hot water with things like WorldEdit commands (e.g. `//wand`). I think you need to match only a single `/`.

```suggestion
            .replaceFirst("^/", ""); // remove a single leading slash
```

John-Paul-R commented on src/main/java/dev/jpcode/kits/command/KitCommandsManagerCommand.java:
```diff
@@ -0,0 +1,99 @@
+        Kit kit = getKit(kitName);
+        if (kit == null) return 0;
+
+        String command = StringArgumentType.getString(context, "command")
```
food for thought, I wonder how `/execute` reads commands with syntax highlighting/suggestions, and if we could use that here. Doesn't have to happen for PR though.

John-Paul-R commented on src/main/java/dev/jpcode/kits/KitsCommandRegistry.java:
```diff
@@ -201,6 +202,27 @@ public static void register(
+                    )
+                )
+                .then(literal("remove")
+                    .then(argument("command", StringArgumentType.greedyString())
```
Can you add this todo? Want to do this eventually, but doesn't have to happen here/now.

```suggestion
                    // TODO: suggestions based on existing commands in kit
                    .then(argument("command", StringArgumentType.greedyString())
```

John-Paul-R replied:
actually, I can just make a github issue, derp: https://github.com/John-Paul-R/kits/issues/18

John-Paul-R commented on src/main/java/dev/jpcode/kits/command/KitCommandsManagerCommand.java:
```diff
@@ -0,0 +1,99 @@
+        String kitName = StringArgumentType.getString(context, "kit_name");
+        ServerCommandSource source = context.getSource();
+        Kit kit = getKit(kitName);
+        if (kit == null) return 0;
```
isn't this case impossible because `getKit` will throw before we get here?

John-Paul-R replied:
```suggestion
```

arnokeesman replied:
I know, editor just got mad there was no return in that case because it doesn't notice the guaranteed throw, will look into cleaning this up

John-Paul-R replied:
You might be able to give it a hint with a `@NotNull` annotation on the `getKit` return type

John-Paul-R commented on src/main/java/dev/jpcode/kits/command/KitCommandsManagerCommand.java:
```diff
@@ -0,0 +1,99 @@
+            if (!existed) throw new CommandException(Text.literal("No command with that index."));
+            saveKit(kitName, kit);
+            source.sendFeedback(() ->
+                    Text.literal(String.format("Removed command \"%s\" from kit '%s'. (click to re-add)", command, kitName))
```
nice.

arnokeesman replied:
Actually, the message in line 80 isn't, used the number/index from `/...list` before to remove commands, never updated that message.
(Changed how that worked so I could use those command suggestions without it messing up after you removed `#1` then wanted to remove `#2` but it moved to `#1` so it'd be a mess without running `/...list` again)

John-Paul-R commented:
I'm a big fan of these changes. Very nice. Big thanks for the PR

arnokeesman committed:
fix the comments from the PR

arnokeesman committed:
suggest commands to delete #18

arnokeesman committed:
starterkit fixes
receiving your starterkit will also execture kit commands now
config option added that sets kit cooldown for starterkits as well (enabled by default)
